### DISCUSSION TITLE
Using the Crazyradio 2.0 with a custom drone and custom receiver hardware

### DISCUSSION THREAD

**Discussion by j-wanner** [bitcraze/discussions#1020] [Q&A] [answered]
Dear Crazyflie team and community,

I would like to evaluate the possibilities to use the Crazyradio 2.0 for enhanced communication with a custom winged drone, in the case of running offboard controls and using motion capture state feedback. As I am not too familiar with the hardware or the Python API, I would like to ask a few questions regarding the feasibility and benefits of such an approach.

1) Which communication protocol is used to transmit commands from the Crazyradio 2.0?
2) Which type of receiver architectures and microprocessors could be used to receive commands from the Crazyradio 2.0 with the lowest possible latency? Would for instance a Seeduino XIAO with WiFi and BLE compatibility using the ESP32 architecute be usable, or must it be an nRF-compatible processor such as the nRF52840 for BLE?
3) What is the typical communication latency over a range of 20 meters that is achieved by the Crazyradio 2.0 and does it remain stable? Are there requirements for the receiver antenna gains? At what range does packet loss become significant?
5) Do you know of other lightweight (< 10 g) commercially available receiver boards that have already been tested to function well with the Crazyradio 2.0 and have a stable communication range of at least 20 meters?
4) Are specific packages or code structures required on the custom receiver side (and must code be written Python or can it be Arduino code) to initiate connection with the Crazyradio and to decode the packets sent using Crazyradio Python API? Are there examples of this code as used on the receiver side?

Thank you in advance, I know these are quite a few questions. I wish to achieve a low and predicatble latency communciation link from an offboard computer running control computations given motion capture feedback to a drone-side receiver architecture that is as light as possible. Perhaps the Crazyradio is a great solution for this.

**Comment by ataffanel** [comment, marked as answer]
Hi, I will try to answer the questions in order:

1. Crazyradio 2 uses a radio protocol compatible with Nordic's Enhanced ShockBurst (ESB) radio protocol. This is because it is currently compatible with Crazyradio PA which uses an nRF24LU1 that has ESB implemented in hardware. This provides bidirectional packet communication with the Crazyflie which is used by the higher-level protocol, CRTP.
2. Crazyradio 2 is physically able to run BLE but it is not implemented yet. BLE has a minimal latency of 7.5ms so we will likely not use it as a main mode of communication with the Crazyflie. When not using BLE any nordic chip can be used as well as any BLE chip with low level access to the radio, unfortunately I do not know of any chip other than Nordic chips that allows low level access to the BLE radio (ESP does not, if you find any please tell me!). The nRF52840 can communicate using Nordic modulation, compatible with all nRF24 and nRF5x, but also give access to all BLE modulations.
3. This is a very hard question because "it depends" is the only possible answer :-). First of all radio communication will depend of the environment, and the GFSK modulation used by nordic chip might not be the most robust, I have seen people use Chirp radios for example for as-close-as-possible-hard-realtime. Now, it does work quite well in not too hard environment. Latency does not depends of the distance (for measurable quantities) and in theory can be very low: since we have low level access to the radio, as soon as the packet has left the transmitter, it is received in the receiver interrupt routine. When adding USB this gives a theoretical minimum latency of 1ms. With Crazyflie we have a roundtrip latency of about 5ms from PC to Crazyradio to nRF51 to STM32 and back. Crazyradio has PA/LNA, Crazyflie has also a PA/LNA but for 20m I do not think it is necessary. I would suggest doing experiments with a nRF5 dev kit.
4. I would expect any nRF5-based modules to work well. The key would be to be able to reprogram the chip.
5. You would need to implement your own receiver. Since we use ESB, you can use the Nordic's provided dev kit (nRF connect) that supports ESB PRX mode, which is what the Crazyflie implements. The nRF5 will likely be written in C. You could look at the crazyflie2-nrf-firmware project for a receiver example but this is not a very modern code. For Crazyradio there is the Python lib, a Rust lib and C++ implementation of the USB protocol to send/receive raw packets.

I hope I could answer your questions well enough. Using Crazyradio with other robot is something I always wanted to make possible and I plan to develop new communication protocols with this in mind, by making libs for both side. In the meantime though, the current ESB implementation could be usable for your use-case. Do not hesitate if you need more information.

**Reply by j-wanner** [reply]
Thanks so much for your detailed response, and apologies for my delay in repyling. Great to know that the nRF52840 chip would be compatible with ESB, and perhaps I'll look into the nRF5 dev kit. It seems that some more extensive firmware development on the receiver side would be required, which is currently out of my project scope as in a first stage I wish to gain some insight on the potential performance of using such an architecture. Is it possible to flash the firmware of the Crazyflie-NRF receiver you mentioned without modifications onto a nRF52840 chip to perform some simple communciation tests with the crazyradio hardware, or will errors be thrown because the Crazyflie sensor connections etc. are missing?
